test/integration/apiserver/flowcontrol/fight_test.go

```diff
@@ -0,0 +1,162 @@
+	"k8s.io/client-go/informers"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/workqueue"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/klog/v2"
+)
+
+const timeFmt = "2006-01-02T15:04:05.999"
+
+func TestConfigConsumerFight(t *testing.T) {
+	// Disable the APF FeatureGate so that the normal config consumer
+	// controller does not interfere
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.APIPriorityAndFairness, false)()
+	_, loopbackConfig, closeFn := setup(t, 100, 100)
+	defer closeFn()
+	ctx := context.Background()
+	const size = 3
+	fsName := fcboot.MandatoryPriorityLevelConfigurationCatchAll.Name
+	stopCh := make(chan struct{})
+	now := time.Now()
+	clk := clock.NewFakeClock(now)
+	ctlrs := map[bool][]utilfc.TestableInterface{
+		false: make([]utilfc.TestableInterface, size),
+		true:  make([]utilfc.TestableInterface, size)}
+	foreach := func(visit func(invert bool, i int, ctlr utilfc.TestableInterface)) {
+		for i := 0; i < size; i++ {
+			// The order of the following iteration is not deterministic,
+			// and that is good.
+			for invert, slice := range ctlrs {
+				visit(invert, i, slice[i])
+			}
+		}
+	}
+	foreach(func(invert bool, i int, _ utilfc.TestableInterface) {
+		myConfig := rest.CopyConfig(loopbackConfig)
+		myConfig = rest.AddUserAgent(myConfig, fmt.Sprintf("invert=%v, i=%d", invert, i))
+		myClientset := clientset.NewForConfigOrDie(myConfig)
+		fcIfc := myClientset.FlowcontrolV1alpha1()
+		// Wait until at least one FlowSchema has been defined by the config producer
+		err := wait.Poll(time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+			_, err := fcIfc.FlowSchemas().Get(ctx, fsName, metav1.GetOptions{})
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		informerFactory := informers.NewSharedInformerFactory(myClientset, 0)
+		fieldMgr := utilfc.ConfigConsumerAsFieldManager
+		foundToDangling := func(found bool) bool { return !found }
+		if invert {
+			fieldMgr = fieldMgr + "x"
+			foundToDangling = func(found bool) bool { return found }
+		}
+		ctlr := utilfc.NewTestable(
+			fmt.Sprintf("Controller%d[invert=%v]", i, invert),
+			clk,
+			func(workqueue.RateLimitingInterface) {},
+			fieldMgr,
+			foundToDangling,
+			informerFactory,
+			fcIfc,
+			200,         // server concurrency limit
+			time.Minute, // request wait limit
+			metrics.PriorityLevelConcurrencyObserverPairGenerator,
+			fqtesting.NewNoRestraintFactory(),
+		)
+		ctlrs[invert][i] = ctlr
+		informerFactory.Start(stopCh)
+		if !ctlr.WaitForCacheSync(stopCh) {
+			t.Fatalf("Never achieved initial sync for i=%d, invert=%v", i, invert)
+		}
+	})
+	var writeCount, countAtStart int
+	var tStart time.Time
+	const testN = 30
+	nextTime := clk.Now()
+	for j := 0; j < 2+testN; j++ {
+		time.Sleep(time.Millisecond * 150)
```

Comment:
why do we need this? could this potentially become flaky in the future?

Reply:
This is to allow certain goroutines --- specifically, the work queue workers in the controllers --- a chance to do their thing.  Yes, in principle it could be flaky.  I do not recall whether CPU overloads are prevented during integration testing.  Does anybody here know?

This is by no means the only test that has this issue.

FYI, the desire to be able to have tests that do not rely on sleeping to get goroutines done is the reason https://github.com/kubernetes/apiserver/blob/master/pkg/util/flowcontrol/fairqueuing/testing/clock/event_clock.go exists.  You see that our QueueSet tests use that.  Unfortunately, using the event clocks for controllers would require some updates to the workqueue package (since workqueue can merge items).

Reply:
Would it be possible to poll until the controller is in the desired state? Otherwise I think this is very likely to be flaky. Yes, we have significant CPU contention in tests.

Reply:
Even running `go test -parallel=100 -count=1000 -run=ConfigConsumerFight` or whatever, although confidence building, often fails to be as harsh as the actual CI test environment.

Reply:
1. I should correct my comment above.  The sleep is to allow the notifications to come in and update the informer local caches; the work queues are deliberately cut out in this test.

2. So yes, I think there is a way to test whether the needed state has been reached.  First, in each iteration, keep track of the last (thankfully the SyncOne calls are serialized!) ResourceVersion returned from each object write.  Then replace the sleep with a test of whether a notification of that ResourceVersion of that object has been received for each controller.  The first iteration will require its own setup, which is do-able.